include/igl/predicates/predicates.cpp

```diff
@@ -0,0 +1,268 @@
+// This file is part of libigl, a simple c++ geometry processing library.
+//
+// Copyright (C) 2019 Qingnan Zhou <qnzhou@gmail.com>
+//
+// This Source Code Form is subject to the terms of the Mozilla Public License 
+// v. 2.0. If a copy of the MPL was not distributed with this file, You can 
+// obtain one at http://mozilla.org/MPL/2.0/.
+#include <igl/predicates/predicates.h>
+#include <predicates.h>
+
+namespace igl {
+namespace predicates {
+
+using REAL = double;
+
+template<typename Vector2D>
+IGL_INLINE Orientation orient2d(
+        const Eigen::MatrixBase<Vector2D>& pa,
+        const Eigen::MatrixBase<Vector2D>& pb,
+        const Eigen::MatrixBase<Vector2D>& pc) {
+
+    using Point = Eigen::Matrix<REAL, Eigen::Dynamic, Eigen::Dynamic>;
+    const Point& a = pa.template cast<REAL>();
```

Comment:
Should we assert that `std::is_same<Vector2d::Scalar, REAL>::value` instead of casting the type of the argument? I wish there was a template trait for detecting narrowing conversion and avoid them, but in the meantime maybe it would be better to prevent automatic cast altogether? (Or simply allow both `float` and `double`, but not `long long double` for example?